### PR TITLE
Sliding buffer function

### PR DESCRIPTION
**Pull request by daywiss**
Generalized version of batch and "window" functionality as discussed in #298

### CODE REVIEW

**Comment by jeromew** [comment]
thanks @daywiss for the PR ; I added some remarks on the PR directly on the code.

**Comment by vqvu** [comment]
Some additional comments:

Some of the tests are redundant. For example, `_.slidingBuffer(2, 1)` and `_.slidingBuffer(3, 1)` aren't really needed if you already have `_.slidingBuffer(4, 1)` and `_.slidingBuffer(1, 1)`. They're not edge cases. Similarly for `_.slidingBuffer(2, 2)` and `_.slidingBuffer(3, 3)`.

The test with no second argument can be dropped. We don't allow optional arguments in transforms.

You can drop much of the "partial application" tests. They're only there to make sure the top-level transform is wired up correctly. You can do the real testing in the main "arraystream" test.

**Comment by daywiss** [comment]
Thanks guys, I'll update when I get a chance

**Comment by daywiss** [comment]
I just want to say to add the "tail" portion of this behavior of scalarx slidingbuffer complicates code since it requires processing and emitting after the stream ends rather than just emitting the current contents of buffer. I also don't feel it adds much, if any practical value.

**Comment by vqvu** [comment]
You're over-complicating things. This modification of your original code passed all the test and is just as simple. The only major addition is the for-loop right before the stream ends.

```javascript
Stream.prototype.slidingBuffer = function (n, m) {
    var windowed = [];
    var missing = 0;

    if (typeof n != 'number' || n <= 0) {
        throw new Error('buffer size must be >0');
    }

    if (typeof m != 'number' || m <= 0) {
        throw new Error('buffer skip must be >0');
    }

    return this.consume(function (err, x, push, next) {
        if (err) {
            push(err);
            next();
        }
        else if (x === nil) {
            while (windowed.length) {
                push(null, windowed);
                windowed = windowed.slice(m);
            }
            push(null, nil);
        }
        else {
            if (missing > 0) {
                missing--;
            }
            else {
                windowed.push(x);
                if (windowed.length === n) {
                    push(null, windowed);
                    missing = Math.max(m - n, 0);
                    windowed = windowed.slice(m);
                }
            }
            next();
        }

    });
};
```

Also, can you change `n` and `m` to more descriptive names? Say `count` and `skip`? `batch` can get away with just using `n` because there's only one argument, and `n` usually means a length.

**Comment by daywiss** [comment]
@vqvu you are right, that is a much simpler solution. I actually tried hard to simplify my code but i had trouble visualizing it. i just tested your solution, id be happy to use that. i still have reservations about this functionality, but maybe you can suggest to me some real world use cases for this type of behavior.

**Comment by vqvu** [comment]
I don't have a real world use case for this behavior, but I also don't have a real world use case for your preferred behavior. As I said, I think your preferred behavior only makes sense in the `skip == 1` case, and it's very easy to get to that behavior using a subsequent filter on the size of the buffer.

The way I think about the behavior of this operator is that you have a window of size `count` which starts at the beginning of the stream and slides by `skip` element each time it emits something new.  The key thing here is where the window starts, not where it ends. It doesn't make sense to abort when the start of the window is still in the stream just because the end of the window slides past the end of the stream.

It would make sense if the operator stops *before* the end of the window slides past the end of the stream (i.e., only emit buffers of size `count`). But, I think we both agree that this is not a good behavior for cases where `skip > 1`.

My other reason for preferring this behavior is that I don't like to have functionality that at first glance looks identical to some functionality from another library but actually has a subtle semantic difference. Or rather, there must be a very good reason for doing so (one example here is `highland#flatMap` vs `RxJS#flatMap`). Granted, this operator does not exist in RxJS right now, but if they ever get around to implementing it, I can guarantee you that it will use the behavior from other Rx implementations, and not the similar-but-slightly-different one that Highland uses.

File: lib/index.js [hunk 1]
```diff
@@ -2340,6 +2340,64 @@ Stream.prototype.batchWithTimeOrCount = function (ms, n) {
...
+            push(null, nil);
+        }
+        else {
+            if (missing > 0) {
+                missing--;
+            }
+            else {
+                windowed.push(x);
+                if (windowed.length === count) {
+                    push(null, windowed);
```
Comment (vqvu) [inline lib/index.js]: I thought about this some more, and I think we should be careful about pushing internal state that we want to use later. We don't want a downstream operator to mess with our state. So this should be changed to

```javascript
var nextWindowed = windowed.slice(skip);
push(null, windowed);
// Turns out the Math#max is not required.
missing = skip - count;
windowed = nextWindowed;
```

Same with the while-loop above.

The concern here is code that looks like this.

```javascript
_([1, 2, 3]).slidingBuffer(3, 1)
    .doTo(function (buf) {
        if (buf.length > 1) {
            buf[1] = 0;
        }
    })
    .toArray(_.log);
// => [[1, 0, 3], [0, 0], [0]]
// => should really be [[1, 0, 3], [2, 0], [3]]
```

Please add a test for this case as well.
